File: server/internal/models/forum.go
package models

import "time"

type ForumCategory string

const (
	CategoryGeneral      ForumCategory = "general"
	CategoryAcademic     ForumCategory = "academic"
	CategoryAssignment   ForumCategory = "assignment"
	CategoryQuestion     ForumCategory = "question"
	CategoryAnnouncement ForumCategory = "announcement"
)

type ForumThread struct {
	ID          int           `json:"id" db:"id"`
	CollegeID   int           `json:"college_id" db:"college_id"`
	CourseID    int           `json:"course_id" db:"course_id"`
	Category    ForumCategory `json:"category" db:"category"`
	Title       string        `json:"title" db:"title"`
	Content     string        `json:"content" db:"content"`
	AuthorID    int           `json:"author_id" db:"author_id"`
	IsPinned    bool          `json:"is_pinned" db:"is_pinned"`
	IsLocked    bool          `json:"is_locked" db:"is_locked"`
	ViewCount   int           `json:"view_count" db:"view_count"`
	ReplyCount  int           `json:"reply_count" db:"reply_count"`
	LastReplyAt *time.Time    `json:"last_reply_at" db:"last_reply_at"`
	LastReplyBy *int          `json:"last_reply_by" db:"last_reply_by"`
	Tags        []string      `json:"tags" db:"tags"`
	CreatedAt   time.Time     `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time     `json:"updated_at" db:"updated_at"`

	// Derived fields
	AuthorName   string `json:"author_name" db:"author_name"`
	AuthorAvatar string `json:"author_avatar" db:"author_avatar"`
	CourseName   string `json:"course_name" db:"course_name"`
}

type ForumReply struct {
	ID               int       `json:"id" db:"id"`
	ThreadID         int       `json:"thread_id" db:"thread_id"`
	ParentID         *int      `json:"parent_id" db:"parent_id"`
	Content          string    `json:"content" db:"content"`
	AuthorID         int       `json:"author_id" db:"author_id"`
	IsAcceptedAnswer bool      `json:"is_accepted_answer" db:"is_accepted_answer"`
	LikeCount        int       `json:"like_count" db:"like_count"`
	CreatedAt        time.Time `json:"created_at" db:"created_at"`
	UpdatedAt        time.Time `json:"updated_at" db:"updated_at"`
	CollegeID        int       `json:"college_id" db:"college_id"`

	// Derived fields
	AuthorName   string `json:"author_name" db:"author_name"`
	AuthorAvatar string `json:"author_avatar" db:"author_avatar"`
	HasLiked     bool   `json:"has_liked" db:"-"`
}

type ForumThreadFilter struct {
	CollegeID  int
	CourseID   *int
	Category   *ForumCategory
	Tag        *string
	AuthorID   *int
	Search     *string
	PinnedOnly *bool
	Limit      int
	Offset     int
}

File: client/src/lib/forum.ts
import type { ForumCategory, ForumReply, ForumThread } from './types';

export type ForumThreadApi = {
  id?: number;
  courseId?: number;
  course_id?: number;
  courseName?: string;
  course_name?: string;
  category?: ForumCategory | string;
  title?: string;
  content?: string;
  authorId?: number;
  author_id?: number;
  authorName?: string;
  author_name?: string;
  authorAvatar?: string;
  author_avatar?: string;
  isPinned?: boolean;
  is_pinned?: boolean;
  isLocked?: boolean;
  is_locked?: boolean;
  viewCount?: number;
  view_count?: number;
  replyCount?: number;
  reply_count?: number;
  lastReplyAt?: string;
  last_reply_at?: string;
  lastReplyBy?: number;
  last_reply_by?: number;
  tags?: string[];
  createdAt?: string;
  created_at?: string;
  updatedAt?: string;
  updated_at?: string;
  collegeId?: number;
  college_id?: number;
};

export type ForumReplyApi = {
  id?: number;
  threadId?: number;
  thread_id?: number;
  parentId?: number;
  parent_id?: number;
  content?: string;
  authorId?: number;
  author_id?: number;
  authorName?: string;
  author_name?: string;
  authorAvatar?: string;
  author_avatar?: string;
  isAcceptedAnswer?: boolean;
  is_accepted_answer?: boolean;
  likeCount?: number;
  like_count?: number;
  hasLiked?: boolean;
  has_liked?: boolean;
  createdAt?: string;
  created_at?: string;
  updatedAt?: string;
  updated_at?: string;
  collegeId?: number;
  college_id?: number;
};

function normalizeCategory(raw: string | ForumCategory | undefined): ForumCategory {
  if (
    raw === 'general' ||
    raw === 'academic' ||
    raw === 'assignment' ||
    raw === 'question' ||
    raw === 'announcement'
  ) {
    return raw;
  }
  return 'general';
}

export function normalizeForumThread(raw: ForumThreadApi): ForumThread {
  const createdAt = raw.createdAt ?? raw.created_at ?? new Date().toISOString();
  const updatedAt = raw.updatedAt ?? raw.updated_at ?? createdAt;
  return {
    id: raw.id ?? 0,
    courseId: raw.courseId ?? raw.course_id ?? 0,
    courseName: raw.courseName ?? raw.course_name,
    category: normalizeCategory(raw.category),
    title: raw.title ?? '',
    content: raw.content ?? '',
    authorId: raw.authorId ?? raw.author_id ?? 0,
    authorName: raw.authorName ?? raw.author_name ?? 'Unknown',
    authorAvatar: raw.authorAvatar ?? raw.author_avatar,
    isPinned: raw.isPinned ?? raw.is_pinned ?? false,
    isLocked: raw.isLocked ?? raw.is_locked ?? false,
    viewCount: raw.viewCount ?? raw.view_count ?? 0,
    replyCount: raw.replyCount ?? raw.reply_count ?? 0,
    lastReplyAt: raw.lastReplyAt ?? raw.last_reply_at,
    lastReplyBy: raw.lastReplyBy ?? raw.last_reply_by,
    tags: Array.isArray(raw.tags) ? raw.tags : [],
    createdAt,
    updatedAt,
    collegeId: raw.collegeId ?? raw.college_id ?? 0,
  };
}

export function normalizeForumReply(raw: ForumReplyApi): ForumReply {
  const createdAt = raw.createdAt ?? raw.created_at ?? new Date().toISOString();
  const updatedAt = raw.updatedAt ?? raw.updated_at ?? createdAt;
  return {
    id: raw.id ?? 0,
    threadId: raw.threadId ?? raw.thread_id ?? 0,
    parentId: raw.parentId ?? raw.parent_id,
    content: raw.content ?? '',
    authorId: raw.authorId ?? raw.author_id ?? 0,
    authorName: raw.authorName ?? raw.author_name ?? 'Unknown',
    authorAvatar: raw.authorAvatar ?? raw.author_avatar,
    isAcceptedAnswer: raw.isAcceptedAnswer ?? raw.is_accepted_answer ?? false,
    likeCount: raw.likeCount ?? raw.like_count ?? 0,
    hasLiked: raw.hasLiked ?? raw.has_liked ?? false,
    createdAt,
    updatedAt,
    collegeId: raw.collegeId ?? raw.college_id ?? 0,
  };
}
